test/gtest/uct/uct_test.cc

```diff
@@ -145,66 +145,136 @@ void uct_test::set_sockaddr_resources(uct_md_h md, char *md_name, cpu_set_t loca
+            fastest_res = const_cast<resource*>(&*res);
+        }
+    } else {
+        resources.push_back(&*res);
+    }
+}
+
+static void commit_tcp_fastest_dev(std::vector<const resource*>& resources,
+                                   resource* &fastest_res)
+{
+    if (fastest_res != NULL) {
+        ucs_assert(fastest_res->tl_name == "tcp");
+        resources.push_back(fastest_res);
+    }
+}
+
 std::vector<const resource*> uct_test::enum_resources(const std::string& tl_name,
-                                                      bool loopback) {
+                                                      bool loopback)
+{
     static std::vector<resource> all_resources;
 
     if (all_resources.empty()) {
+        uct_iface_params_t params;
         uct_md_resource_desc_t *md_resources;
         unsigned num_md_resources;
         uct_tl_resource_desc_t *tl_resources;
         unsigned num_tl_resources;
         ucs_status_t status;
+        ucs_async_context_t *async;
+        uct_worker_h worker;
+
+        status = ucs_async_context_create(UCS_ASYNC_MODE_THREAD_SPINLOCK, &async);
+        ASSERT_UCS_OK(status);
+
+        status = uct_worker_create(async, UCS_THREAD_MODE_SINGLE, &worker);
+        ASSERT_UCS_OK(status);
 
         status = uct_query_md_resources(&md_resources, &num_md_resources);
         ASSERT_UCS_OK(status);
 
+        params.field_mask = UCT_IFACE_PARAM_FIELD_OPEN_MODE |
+                            UCT_IFACE_PARAM_FIELD_DEVICE;
+        params.open_mode  = UCT_IFACE_OPEN_MODE_DEVICE;
+
         for (unsigned i = 0; i < num_md_resources; ++i) {
-            uct_md_h pd;
+            uct_md_h md;
             uct_md_config_t *md_config;
             status = uct_md_config_read(md_resources[i].md_name, NULL, NULL,
                                         &md_config);
             ASSERT_UCS_OK(status);
 
             {
                 scoped_log_handler slh(hide_errors_logger);
-                status = uct_md_open(md_resources[i].md_name, md_config, &pd);
+                status = uct_md_open(md_resources[i].md_name, md_config, &md);
             }
             uct_config_release(md_config);
             if (status != UCS_OK) {
                 continue;
             }
 
             uct_md_attr_t md_attr;
-            status = uct_md_query(pd, &md_attr);
+            status = uct_md_query(md, &md_attr);
             ASSERT_UCS_OK(status);
 
-            status = uct_md_query_tl_resources(pd, &tl_resources, &num_tl_resources);
+            status = uct_md_query_tl_resources(md, &tl_resources, &num_tl_resources);
             ASSERT_UCS_OK(status);
 
             for (unsigned j = 0; j < num_tl_resources; ++j) {
+                uct_iface_config_t *iface_config;
+                uct_iface_attr_t iface_attr;
+                uct_iface_h iface;
                 resource rsc;
+
+                status = uct_md_iface_config_read(md, tl_resources[j].tl_name,
```

Comment:
can we do this only if tl_resources[j].tl_name is tcp?

Reply:
yes, we can do - done